test-loop-tests/src/setup/builder.rs

```diff
@@ -467,40 +477,63 @@ impl TestLoopBuilder {
         self
     }
 
+    fn ensure_tempdir(mut self) -> Self {
+        self.test_loop_data_dir.get_or_insert_with(|| tempfile::tempdir().unwrap());
+        self
+    }
+
     fn build_impl(mut self) -> TestLoopEnv {
-        let mut datas = Vec::new();
-        let tempdir =
-            self.test_loop_data_dir.take().unwrap_or_else(|| tempfile::tempdir().unwrap());
-        let network_shared_state = TestLoopNetworkSharedState::new();
-        for idx in 0..self.clients.len() {
-            let account = &self.clients[idx];
-            let is_archival = self.archival_clients.contains(account);
-            let data = self.setup_client(idx, &tempdir, is_archival, &network_shared_state);
-            datas.push(data);
-        }
+        let warmup_pending = self.warmup_pending.clone();
+        self.test_loop.send_adhoc_event("warmup_pending".into(), move |_| {
+            assert!(
+                !warmup_pending.load(Ordering::Relaxed),
+                "Warmup is pending! Call env.warmup() or builder.skip_warmup()"
+            );
+        });
 
-        let env = TestLoopEnv { test_loop: self.test_loop, datas, tempdir };
-        if self.warmup { env.warmup() } else { env }
```

Comment:
Moved this outside environment builder as it's a pain to have it here.. This restricts doing anything else with the environment after building.

Ideally this function belongs to `env` and we should treat it that way, not internally hidden in `builder.build()` implementation.

I've designed for a fool proof way to ensure that `warmup()` is called every time and not missed. We send an adhoc event checking if a shared valiable `warmup_pending` has been reset or not